sockets/sockets.go

```diff
@@ -24,6 +25,10 @@ func ConfigureTransport(tr *http.Transport, proto, addr string) error {
 		return configureNpipeTransport(tr, proto, addr)
 	default:
 		tr.Proxy = http.ProxyFromEnvironment
+		tr.DisableCompression = false
```

Comment:
Still looking at this one, but I don't have a good answer. The general issue here is that as you described in the commit message (e.g.) `configureUnixTransport` mutates this option https://github.com/docker/go-connections/blob/5cc4da5c08cd0df3e0a45da21ebd07e131109bd2/sockets/sockets_unix.go#L19-L24, which means it's now ambiguous if it was the caller's _intent_ to have compression disabled, or if it was a side-effect of `configureUnixTransport` having been called before we reach this (and have implicitly mutated that property).

Effectively, that could also be considered a bug in the caller code, i.e., the caller code has re-used a transport, and tried to configure it multiple times. However, I don't know if there's code-paths where this happens implicitly; do you know if those are present?

If we unconditionally enable compression here, that would also mean that there's no option for a HTTP transport to disable compression, which _could be_ a legitimate case, and may also be unexpected 🤔

But honestly, I have no good suggestion what would be best here (other than suggesting; don't re-use transport (or more factually: don't call `ConfigureTransport` on the same transport multiple times) if it should not be re-used.

Reply:
I guess the alternative could be to remove that part from `configureUnixTransport`, and somehow make sure that when the transport is _used_, the compression is disabled, but that would require changes elsewhere to keep the existing behavior.

Reply:
> Effectively, that could also be considered a bug in the caller code, i.e., the caller code has re-used a transport, and tried to configure it multiple times. However, I don't know if there's code-paths where this happens implicitly; do you know if those are present?

Yeah, there's one major user of that pattern: the official Docker client :see_no_evil: 

- `defaultHTTPClient` initializes a new transport: https://github.com/moby/moby/blob/e9efc0a361de39903d47fe12f628c92fc095397b/client/client.go#L237-L238
- Then, `WithHost` calls `ConfigureTransport` on that same transport: https://github.com/moby/moby/blob/e9efc0a361de39903d47fe12f628c92fc095397b/client/options.go#L71-L73
- Both are called from `NewClientWithOpts`: https://github.com/moby/moby/blob/e9efc0a361de39903d47fe12f628c92fc095397b/client/client.go#L181-L197

I think we can't / shouldn't create a new transport in `WithHost` (instead of mutating the existing one), as there're at least two other option funcs that mutate the same transport:

- `WithTLSClientConfig`: https://github.com/moby/moby/blob/e9efc0a361de39903d47fe12f628c92fc095397b/client/options.go#L136-L155
- `WithDialContext`: https://github.com/moby/moby/blob/e9efc0a361de39903d47fe12f628c92fc095397b/client/options.go#L50-L58

> That part was added in https://github.com/docker/go-connections/commit/88d5fb2e848739e150018fa071fa17a8d25f4395, but it's possible that code was moved from elsewhere 🤔

It was! It comes from this commit [`moby/moby@fb7ceeb1`](https://github.com/moby/moby/commit/fb7ceeb170447cdbef3e74a92358eb9b084c0bf7). I think it makes a lot of sense to disable compression on UNIX socket, as it's only going to make the CPU spin needlessly more.

IMO it's fine to reconfigure compression when the transport protocol is changed as it's somewhat tied to it, ie. you generally want compression for HTTP over TCP, and you don't want it for HTTP over UNIX / npipe sockets.

Currently, and presumably since forever, compression has been disabled for HTTP over TCP due to the transport being initialized by `defaultHTTPClient` with a UNIX / npipe transport.

> which means it's now ambiguous if it was the caller's intent to have compression disabled, or if it was a side-effect of configureUnixTransport having been called before we reach this (and have implicitly mutated that property).

I think it's OK to warn `ConfigureTransport` and `WithHost` callers through a doc comment that both functions will toggle compression based on what protocol is passed. If users really want to enforce that value, they should set it _after_ these functions are called.

Reply:
If we agree on that :arrow_up:, I need to update the doc comment on `ConfigureTransport` (and on `WithHost` in moby/moby).

Reply:
> Yeah, there's one major user of that pattern: the official Docker client 🙈

Yes, that's what I was afraid of :disappointed:

> I think it makes a lot of sense to disable compression on UNIX socket, as it's only going to make the CPU spin needlessly more.

Yup, I definitely agree with that; it _does_ make sense to not have compression there. The tricky bit is indeed that (per the other comment) changing that option also affects other paths :disappointed:

> IMO it's fine to reconfigure compression when the transport protocol is changed as it's somewhat tied to it, ie. you generally want compression for HTTP over TCP, and you don't want it for HTTP over UNIX / npipe sockets.

It's probably fine to keep the status quo for now (at least until we cleaned up all of this). I had feature requests, such as https://github.com/moby/moby/issues/1266 in mind, where it may be desirable to disable compression (for local connections). Admitted; in that case it's mostly about compressing the layers (which would have a much bigger impact for those cases)

> I think it's OK to warn ConfigureTransport and WithHost callers through a doc comment that both functions will toggle compression based on what protocol is passed. If users really want to enforce that value, they should set it after these functions are called.

Documenting definitely won't hurt. At least we shouldn't try to hide that information.


Perhaps (in future) we should look at making this all more atomic somehow (but we'd have to look what the best approach is)

Reply:
After discussing privately this change with @thaJeztah, I'll move this change to another commit/PR to make it clear why it's changed.